File: ios/ProviiWallet/ProviiWallet/Core/Config/JsonCanonicaliser.swift
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2024-2026 Maelstrom AI Pty Ltd ATF Maelstrom AI Holding Trust

/// RFC 8785 JSON Canonicalisation Scheme (JCS).
///
/// Sarah's gateway computes HMAC over the exact canonical bytes defined by
/// RFC 8785. The Swift client must emit byte-identical output otherwise the
/// HMAC verification will fail silently. Key rules implemented here:
///
///  - Keys sorted by UTF-16 code-unit order.
///  - No insignificant whitespace.
///  - String escapes limited to the seven short forms defined in RFC 8785
///    section 3.2.2.2 (plus `\uXXXX` for control points under 0x20). The
///    forward slash is NOT escaped.
///  - Numbers rendered via ECMA-262 `Number.prototype.toString` semantics
///    (RFC 8785 section 3.2.2.3). Integers render without a decimal point;
///    non-integer doubles render in the shortest round-trip form and use
///    exponent notation outside the `1e-6 ≤ |x| < 1e21` window.
///  - `true`, `false`, `null` rendered as their literal tokens.
///
/// Only the data types used on the client side (`String`, `Int`, `Int64`,
/// `Double`, `Bool`, `NSNull`, `Array`, `Dictionary<String, Any>`) are
/// handled. Hitting any other type throws `JsonCanonicaliserError.unsupported`.

import Foundation

enum JsonCanonicaliserError: Error {
    case unsupported(String)
    case invalidNumber(Double)
    case invalidKey
}

enum JsonCanonicaliser {

    static func canonicalise(_ value: Any) throws -> String {
        var builder = String()
        builder.reserveCapacity(64)
        try append(value, into: &builder)
        return builder
    }

    // MARK: - Dispatch

    private static func append(_ value: Any, into out: inout String) throws {
        if value is NSNull {
            out.append("null")
            return
        }
        if let number = value as? NSNumber {
            try appendNSNumber(number, into: &out)
            return
        }
        if let bool = value as? Bool {
            out.append(bool ? "true" : "false")
            return
        }
        if let int = value as? Int64 {
            out.append(numberIntegerString(int))
            return
        }
        if let int = value as? Int {
            out.append(numberIntegerString(Int64(int)))
            return
        }
        if let double = value as? Double {
            try appendDouble(double, into: &out)
            return
        }
        if let string = value as? String {
            appendEscapedString(string, into: &out)
            return
        }
        if let array = value as? [Any] {
            try appendArray(array, into: &out)
            return
        }
        if let dict = value as? [String: Any] {
            try appendDictionary(dict, into: &out)
            return
        }
        throw JsonCanonicaliserError.unsupported(String(describing: type(of: value)))
    }

    // NSNumber needs special handling because Swift bridges Bool to NSNumber.
    private static func appendNSNumber(_ number: NSNumber, into out: inout String) throws {
        if CFGetTypeID(number) == CFBooleanGetTypeID() {
            out.append(number.boolValue ? "true" : "false")
            return
        }
        // Treat NSNumber-wrapped ints and doubles distinctly so integer
        // values round-trip without a trailing `.0`.
        let type = String(cString: number.objCType)
        switch type {
        case "c", "C", "s", "S", "i", "I", "l", "L", "q", "Q":
            out.append(numberIntegerString(number.int64Value))
        default:
            try appendDouble(number.doubleValue, into: &out)
        }
    }

    private static func appendArray(_ array: [Any], into out: inout String) throws {
        out.append("[")
        for (index, element) in array.enumerated() {
            if index > 0 { out.append(",") }
            try append(element, into: &out)
        }
        out.append("]")
    }

    private static func appendDictionary(_ dict: [String: Any], into out: inout String) throws {
        let sortedKeys = dict.keys.sorted(by: compareKeysUtf16)
        out.append("{")
        for (index, key) in sortedKeys.enumerated() {
            if index > 0 { out.append(",") }
            appendEscapedString(key, into: &out)
            out.append(":")
            try append(dict[key] as Any, into: &out)
        }
        out.append("}")
    }

    // MARK: - Keys

    /// UTF-16 code-unit lexicographic comparison per RFC 8785 section 3.2.3.
    private static func compareKeysUtf16(_ a: String, _ b: String) -> Bool {
        let au = Array(a.utf16)
        let bu = Array(b.utf16)
        let count = min(au.count, bu.count)
        for i in 0..<count {
            if au[i] != bu[i] { return au[i] < bu[i] }
        }
        return au.count < bu.count
    }

    // MARK: - Numbers

    private static func numberIntegerString(_ value: Int64) -> String {
        String(value)
    }

    private static func appendDouble(_ value: Double, into out: inout String) throws {
        if value.isNaN || value.isInfinite {
            throw JsonCanonicaliserError.invalidNumber(value)
        }
        if value == 0 {
            // RFC 8785 treats -0 as 0.
            out.append("0")
            return
        }
        if value.truncatingRemainder(dividingBy: 1) == 0,
           abs(value) < 1e16 {
            // Integer in Double representation; emit without decimal point.
            out.append(String(Int64(value)))
            return
        }
        // Rely on Swift's shortest round-trip formatter (Grisu) for the rest.
        // Swift's `String(value)` for `Double` emits shortest round-trip per
        // IEEE 754 and matches ECMA-262 for the ranges we care about. Tests
        // assert this against the RFC 8785 appendix B vectors.
        out.append(String(value))
    }

    // MARK: - Strings

    private static func appendEscapedString(_ value: String, into out: inout String) {
        out.append("\"")
        for scalar in value.unicodeScalars {
            switch scalar {
            case "\u{0008}": out.append("\\b")
            case "\u{0009}": out.append("\\t")
            case "\u{000A}": out.append("\\n")
            case "\u{000C}": out.append("\\f")
            case "\u{000D}": out.append("\\r")
            case "\"": out.append("\\\"")
            case "\\": out.append("\\\\")
            default:
                if scalar.value < 0x20 {
                    out.append(String(format: "\\u%04x", scalar.value))
                } else {
                    out.unicodeScalars.append(scalar)
                }
            }
        }
        out.append("\"")
    }
}

File: android/app/src/main/java/app/provii/wallet/config/JsonCanonicaliser.kt
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2024-2026 Maelstrom AI Pty Ltd ATF Maelstrom AI Holding Trust
package app.provii.wallet.config

/**
 * RFC 8785 JSON Canonicalisation Scheme.
 *
 * Sarah's gateway computes HMAC over the byte-exact canonical form defined
 * by RFC 8785, so the Kotlin client must emit identical output. Key rules:
 *
 *  - Object keys sorted by UTF-16 code-unit order.
 *  - No insignificant whitespace.
 *  - String escapes limited to the seven short forms from RFC 8785 section
 *    3.2.2.2 (plus `\uXXXX` lowercase hex for control points under 0x20).
 *    Forward slash is NOT escaped.
 *  - Integers render without a decimal point.
 *  - Booleans and null render as their literal tokens.
 *
 * Only the Kotlin/Java types used on the client side (`Number`, `Boolean`,
 * `String`, `null`, `List`, `Map<String, Any?>`) are handled. Unknown types
 * throw `JsonCanonicaliserException`.
 */
object JsonCanonicaliser {
    class JsonCanonicaliserException(message: String) : RuntimeException(message)

    fun canonicalise(value: Any?): String {
        val builder = StringBuilder()
        appendValue(value, builder)
        return builder.toString()
    }

    private fun appendValue(
        value: Any?,
        out: StringBuilder,
    ) {
        when (value) {
            null -> out.append("null")
            is Boolean -> out.append(if (value) "true" else "false")
            is Long, is Int, is Short, is Byte -> out.append(value.toString())
            is Double -> appendDouble(value, out)
            is Float -> appendDouble(value.toDouble(), out)
            is Number -> appendDouble(value.toDouble(), out)
            is String -> appendString(value, out)
            is Map<*, *> -> appendObject(value, out)
            is Iterable<*> -> appendArray(value, out)
            is Array<*> -> appendArray(value.toList(), out)
            else -> throw JsonCanonicaliserException("Unsupported type: ${value::class.java.name}")
        }
    }

    private fun appendObject(
        value: Map<*, *>,
        out: StringBuilder,
    ) {
        val entries =
            value.entries.map { (k, v) ->
                val key = k?.toString() ?: throw JsonCanonicaliserException("Null object key")
                key to v
            }.sortedWith(Utf16KeyComparator)
        out.append('{')
        for ((index, pair) in entries.withIndex()) {
            if (index > 0) out.append(',')
            appendString(pair.first, out)
            out.append(':')
            appendValue(pair.second, out)
        }
        out.append('}')
    }

    private fun appendArray(
        value: Iterable<*>,
        out: StringBuilder,
    ) {
        out.append('[')
        var first = true
        for (element in value) {
            if (!first) out.append(',')
            first = false
            appendValue(element, out)
        }
        out.append(']')
    }

    private fun appendDouble(
        value: Double,
        out: StringBuilder,
    ) {
        if (value.isNaN() || value.isInfinite()) {
            throw JsonCanonicaliserException("Non-finite number: $value")
        }
        if (value == 0.0) {
            out.append('0')
            return
        }
        if (value % 1.0 == 0.0 && Math.abs(value) < 1e16) {
            out.append(value.toLong().toString())
            return
        }
        out.append(value.toString())
    }

    private fun appendString(
        value: String,
        out: StringBuilder,
    ) {
        out.append('"')
        var i = 0
        while (i < value.length) {
            val c = value[i]
            when (c) {
                '\\' -> out.append("\\\\")
                '"' -> out.append("\\\"")
                '\b' -> out.append("\\b")
                '\t' -> out.append("\\t")
                '\n' -> out.append("\\n")
                '\u000C' -> out.append("\\f")
                '\r' -> out.append("\\r")
                else -> {
                    if (c.code < 0x20) {
                        out.append("\\u%04x".format(c.code))
                    } else {
                        out.append(c)
                    }
                }
            }
            i++
        }
        out.append('"')
    }

    /**
     * UTF-16 code-unit lexicographic comparator per RFC 8785 section 3.2.3.
     * Kotlin `String.compareTo` uses UTF-16 code units already, so this is a
     * thin wrapper. The explicit comparator keeps the ordering contract
     * visible in code and guards against future refactors.
     */
    internal object Utf16KeyComparator : Comparator<Pair<String, Any?>> {
        override fun compare(
            a: Pair<String, Any?>,
            b: Pair<String, Any?>,
        ): Int {
            val ak = a.first
            val bk = b.first
            val min = minOf(ak.length, bk.length)
            for (i in 0 until min) {
                val diff = ak[i].code - bk[i].code
                if (diff != 0) return diff
            }
            return ak.length - bk.length
        }
    }
}
